rs/embedders/benches/heap.rs

```diff
@@ -72,57 +90,73 @@ fn throughput(size: Size, step: Step) -> Option<Throughput> {
 fn setup_action(src: Src) -> SetupAction {
     match src {
         Src::Checkpoint => SetupAction::PerformCheckpoint,
+        Src::Mix => SetupAction::PerformCheckpointCallSetup,
         Src::PageDelta | Src::NewAllocation => SetupAction::None,
     }
 }
 
-fn heap_op(op: Op) -> String {
-    match op {
-        Op::Read => "(drop (i64.load (local.get $address)))".into(),
-        Op::Write => "(i64.store (local.get $address) (global.get $counter))".into(),
+fn heap_op(op: Op, mem: Mem) -> String {
+    match (op, mem) {
+        (Op::Read, Mem::Wasm32) => {
+            "(global.set $data (i64.load8_u (i32.wrap_i64 (local.get $address))))"
+        }
+        (Op::Read, Mem::Wasm64) => "(global.set $data (i64.load8_u (local.get $address)))",
+        (Op::Write, Mem::Wasm32) => {
+            "(i64.store8 (i32.wrap_i64 (local.get $address)) (global.get $counter))"
+        }
+        (Op::Write, Mem::Wasm64) => "(i64.store8 (local.get $address) (global.get $counter))",
     }
+    .into()
 }
 
 fn heap_func_init_body(mem: Mem, size: Size, src: Src) -> String {
     let wasm_pages = (size as usize).div_ceil(WASM_PAGE_SIZE_IN_BYTES);
     match src {
-        Src::Checkpoint | Src::PageDelta => String::new(),
+        Src::Checkpoint | Src::PageDelta | Src::Mix => String::new(),
         Src::NewAllocation => format!("(drop (memory.grow (i{mem}.const {wasm_pages})))"),
     }
 }
 
-fn loop_body(op: &str, mem: Mem, dir: Dir, size: Size, step: Step) -> String {
-    let size = size as usize;
+fn loop_body(op: &str, dir: Dir, offset: usize, size: Size, step: Step) -> String {
     let step = step as usize;
     match dir {
-        Dir::Forward => format!(
-            r#"
-                (local.set $address (i{mem}.const 0))
+        Dir::Fwd => {
+            let end = size as usize - 1;
+            format!(
+                r#"
+                (local.set $address (i64.const {offset}))
                 (loop $loop
                     {op}
-                    (local.set $address (i{mem}.add (local.get $address) (i{mem}.const {step})))
-                    (br_if $loop (i{mem}.lt_u (local.get $address) (i{mem}.const {size})))
+                    (local.set $address (i64.add (local.get $address) (i64.const {step})))
+                    (br_if $loop (i64.le_s (local.get $address) (i64.const {end})))
                 )
             "#
-        ),
-        Dir::Backward => format!(
-            r#"
-                (local.set $address (i{mem}.const {size}))
+            )
+        }
+        Dir::Bwd => {
+            let end = size as usize - 1 - offset;
```

Comment:
```suggestion
            let end = size as usize - offset;
```

Reply:
The range of valid addresses is [0..size -1] inclusive, so it looks correct to me...

Reply:
The last step of the forward direction is size - offset so this should be the first step of the backwards direction.

Reply:
The `size` is in bytes, the offset is in bytes as well. Maybe that's the source of confusion?

The simplest example is for size = 4096 bytes the valid addresses are [0..4096 - 1] inclusive, as the address 4096 is out of range.

Reply:
> The last step of the forward direction is size - step
Or maybe this is confusing... The last operation for the forward direction might happen at the `size - 1` address; it depends on the `size` and `step`. The backward direction is "mirrored", so it starts at `size - 1 - offset` and goes down to address `0`.